File: src/content/docs/topics/cs/coding-problems/graphs/417-pacific-atlantic-water-flow-approach2.py
def pacific_atlantic(heights):
    if not heights:
        return []
    rows, cols = len(heights), len(heights[0])
    pac = set()
    atl = set()

    def dfs(r, c, visited):
        if (r, c) in visited:
            return
        visited.add((r, c))
        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and heights[nr][nc] >= heights[r][c]:
                dfs(nr, nc, visited)

    for c in range(cols):
        dfs(0, c, pac)
        dfs(rows - 1, c, atl)
    for r in range(rows):
        dfs(r, 0, pac)
        dfs(r, cols - 1, atl)

    return sorted([r, c] for (r, c) in pac & atl)

h1 = [[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]
assert pacific_atlantic(h1) == [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]
assert pacific_atlantic([[5]]) == [[0, 0]]
h2 = [[1, 1], [1, 1]]
assert sorted(pacific_atlantic(h2)) == [[0,0],[0,1],[1,0],[1,1]]
h3 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
r3 = pacific_atlantic(h3)
assert [2, 2] in r3
assert pacific_atlantic([]) == []
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/graphs/417-pacific-atlantic-water-flow-approach2.go
package main

import (
	"fmt"
	"sort"
)

func pacificAtlantic(heights [][]int) [][]int {
	if len(heights) == 0 {
		return nil
	}
	rows, cols := len(heights), len(heights[0])
	pac := make([][]bool, rows)
	atl := make([][]bool, rows)
	for i := range pac {
		pac[i] = make([]bool, cols)
		atl[i] = make([]bool, cols)
	}
	dirs := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	var dfs func(r, c int, visited [][]bool)
	dfs = func(r, c int, visited [][]bool) {
		if visited[r][c] {
			return
		}
		visited[r][c] = true
		for _, d := range dirs {
			nr, nc := r+d[0], c+d[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols && heights[nr][nc] >= heights[r][c] {
				dfs(nr, nc, visited)
			}
		}
	}

	for c := 0; c < cols; c++ {
		dfs(0, c, pac)
		dfs(rows-1, c, atl)
	}
	for r := 0; r < rows; r++ {
		dfs(r, 0, pac)
		dfs(r, cols-1, atl)
	}

	var result [][]int
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if pac[r][c] && atl[r][c] {
				result = append(result, []int{r, c})
			}
		}
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i][0] != result[j][0] {
			return result[i][0] < result[j][0]
		}
		return result[i][1] < result[j][1]
	})
	return result
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func eqResult(a, b [][]int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i][0] != b[i][0] || a[i][1] != b[i][1] {
			return false
		}
	}
	return true
}

func runTests() {
	h1 := [][]int{{1, 2, 2, 3, 5}, {3, 2, 3, 4, 4}, {2, 4, 5, 3, 1}, {6, 7, 1, 4, 5}, {5, 1, 1, 2, 4}}
	expected1 := [][]int{{0, 4}, {1, 3}, {1, 4}, {2, 2}, {3, 0}, {3, 1}, {4, 0}}
	assert(eqResult(pacificAtlantic(h1), expected1))
	assert(eqResult(pacificAtlantic([][]int{{5}}), [][]int{{0, 0}}))
	h2 := [][]int{{1, 1}, {1, 1}}
	assert(eqResult(pacificAtlantic(h2), [][]int{{0, 0}, {0, 1}, {1, 0}, {1, 1}}))
	assert(len(pacificAtlantic([][]int{})) == 0)
	fmt.Println("all tests pass")
}

func main() { runTests() }
